### PR TITLE
Fuse SwiGLU silu(gate) * up into a Triton autograd op

### PR DESCRIPTION
**Pull request by MasterJH5574**
Eager PyTorch implements `silu(gate) * up` as separate element-wise kernels and saves three activations (`sigmoid(gate)`, `silu(gate)`, `up`) for backward. Since the Qwen3 and DeepSeek-V2-Lite MLP/expert paths are not wrapped in torch.compile, nothing fuses these ops at runtime and the extra `silu(gate)` tensor is wasted activation memory.

Add `pithtrain/operators/silu_mul.py` with a custom autograd Function backed by two Triton kernels: the forward does `silu(gate) * up` in one pass, and the backward recomputes `silu` / `sigmoid` on the fly and writes both `grad_gate` and `grad_up` from a single kernel. Only `gate` and `up` are saved for backward, dropping one activation tensor per MLP/expert layer.

Wire the op into `Qwen3MoeMLP` / `Qwen3MoeExperts` and the DeepSeek-V2- Lite MLP/experts counterparts, replacing the `nn.SiLU()` + elementwise multiply. Add correctness tests against `F.silu(gate) * up` across bf16/fp32 and several shapes including non-BLOCK_SIZE-aligned ones.

### CODE REVIEW

File: pithtrain/operators/silu_mul.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        assert gate.shape == up.shape, f"shape mismatch: {gate.shape} vs {up.shape}"
+        assert gate.dtype == up.dtype, f"dtype mismatch: {gate.dtype} vs {up.dtype}"
+        assert gate.is_contiguous(), "gate must be contiguous"
+        assert up.is_contiguous(), "up must be contiguous"
+        ctx.save_for_backward(gate, up)
+        out = torch.empty_like(gate)
+        n = gate.numel()
+        grid = (triton.cdiv(n, _BLOCK_SIZE),)
+        _silu_mul_fwd_kernel[grid](gate, up, out, n, BLOCK_SIZE=_BLOCK_SIZE)
+        return out
```
[the remark this answers is not in the document]
Comment (MasterJH5574) [reply pithtrain/operators/silu_mul.py:107]: We only support CUDA.

File: pithtrain/operators/silu_mul.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    silu = gate_f * sig
+
+    grad_up = grad_out_f * silu
+    grad_gate = grad_out_f * up_f * sig * (1.0 + gate_f - silu)
+
+    tl.store(grad_gate_ptr + offs, grad_gate.to(gate.dtype), mask=mask)
+    tl.store(grad_up_ptr + offs, grad_up.to(up.dtype), mask=mask)
+
+
+_BLOCK_SIZE = 1024
```
Comment (haok1402) [inline pithtrain/operators/silu_mul.py:90]: @MasterJH5574 nit: Could we make `BLOCK_SIZE` a class-level constant on `_SiLUMul` instead of a module-level `_BLOCK_SIZE`? Having the tuning parameter live in the global scope while the Triton kernels take it as a parameter feels a bit disconnected.

**Review by haok1402 — APPROVED** [review]

**Comment by haok1402** [comment]
@MasterJH5574 Thanks for the proposal! Below are the benchmark throughput on 2xH100 training Qwen3-30B-A3B with 1024 global batch size and 2048 sequence length.  We see a 6% throughput improvement with 1.3 GiB reduction in peak allocated GPU memory.

```
haok@orchard-flame-31:~/Pith-Train-Benchmark$ python3 scripts/pithtrain/benchmark/report.py 
step-time (s), last 7                                                                 Steps        Avg    Med    Min    Max    StdDev
------------------------------------------------------------------------------------  -----------  -----  -----  -----  -----  ------
benchmark/qwen3-30b-a3b/2x8-h100/2pp-1dp-1cp-8ep/1024x2048-bf16/pithtrain_a2afc4.log  9-15 (7/15)  21.31  21.35  20.55  22.41  0.65  
benchmark/qwen3-30b-a3b/2x8-h100/2pp-1dp-1cp-8ep/1024x2048-bf16/pithtrain_bf3a1b.log  9-15 (7/15)  20.06  20.06  20.03  20.10  0.03  
```

---

Before the change,

```
2026-04-14 12:05:38 | INFO | step 00000011/00000015 | step-time 21.352 sec | cross-entropy-loss 2.6574 | learning-rate 1.000000e-06 | gradient-norm 389.9648 | tokens-per-second 98,220 | peak-gpu-memory 59.84 GB
2026-04-14 12:06:01 | INFO | step 00000012/00000015 | step-time 22.407 sec | cross-entropy-loss 2.6251 | learning-rate 1.000000e-06 | gradient-norm 442.8093 | tokens-per-second 93,593 | peak-gpu-memory 59.87 GB
2026-04-14 12:06:23 | INFO | step 00000013/00000015 | step-time 21.535 sec | cross-entropy-loss 2.6408 | learning-rate 1.000000e-06 | gradient-norm 375.8546 | tokens-per-second 97,382 | peak-gpu-memory 59.92 GB
2026-04-14 12:06:44 | INFO | step 00000014/00000015 | step-time 20.555 sec | cross-entropy-loss 2.6304 | learning-rate 1.000000e-06 | gradient-norm 683.4340 | tokens-per-second 102,024 | peak-gpu-memory 59.85 GB
2026-04-14 12:07:05 | INFO | step 00000015/00000015 | step-time 20.595 sec | cross-entropy-loss 2.6812 | learning-rate 1.000000e-06 | gradient-norm 350.7966 | tokens-per-second 101,830 | peak-gpu-memory 59.86 GB
```

After the change,

```
2026-04-14 12:18:25 | INFO | step 00000011/00000015 | step-time 20.096 sec | cross-entropy-loss 2.6573 | learning-rate 1.000000e-06 | gradient-norm 520.0389 | tokens-per-second 104,355 | peak-gpu-memory 58.49 GB
2026-04-14 12:18:45 | INFO | step 00000012/00000015 | step-time 20.039 sec | cross-entropy-loss 2.6249 | learning-rate 1.000000e-06 | gradient-norm 424.4496 | tokens-per-second 104,652 | peak-gpu-memory 58.48 GB
2026-04-14 12:19:05 | INFO | step 00000013/00000015 | step-time 20.031 sec | cross-entropy-loss 2.6407 | learning-rate 1.000000e-06 | gradient-norm 566.5081 | tokens-per-second 104,694 | peak-gpu-memory 58.46 GB
2026-04-14 12:19:26 | INFO | step 00000014/00000015 | step-time 20.056 sec | cross-entropy-loss 2.6300 | learning-rate 1.000000e-06 | gradient-norm 890.4203 | tokens-per-second 104,564 | peak-gpu-memory 58.43 GB
2026-04-14 12:19:46 | INFO | step 00000015/00000015 | step-time 20.029 sec | cross-entropy-loss 2.6809 | learning-rate 1.000000e-06 | gradient-norm 359.0547 | tokens-per-second 104,706 | peak-gpu-memory 58.45 GB
```

File: pithtrain/operators/silu_mul.py (hunk 2 above)
Comment (MasterJH5574) [reply pithtrain/operators/silu_mul.py:90]: Great suggestion.  Maybe do that in some follow-up.
